src/rust/engine/options/src/config.rs

```diff
@@ -1,18 +1,98 @@
 // Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
 // Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-use std::collections::HashSet;
+use std::collections::{HashMap, HashSet};
 use std::fs;
 use std::mem;
 use std::path::Path;
 
+use lazy_static::lazy_static;
+use regex::Regex;
 use toml::value::Table;
 use toml::Value;
 
 use super::id::{NameTransform, OptionId};
 use super::parse::parse_string_list;
 use super::{ListEdit, ListEditAction, OptionsSource};
 
+type InterpolationMap = HashMap<String, String>;
+
+lazy_static! {
+    static ref PLACEHOLDER_RE: Regex = Regex::new(r"%\(([a-zA-Z0-9_.]+)\)s").unwrap();
+}
+
+pub(crate) fn interpolate_string(
+    value: String,
+    replacements: &InterpolationMap,
+) -> Result<String, String> {
+    let caps_vec: Vec<_> = PLACEHOLDER_RE.captures_iter(&value).collect();
+    if caps_vec.is_empty() {
+        return Ok(value);
+    }
+
+    let mut new_value = String::with_capacity(value.len());
+    let mut last_match = 0;
+    for caps in caps_vec {
+        let m = caps.get(0).unwrap();
+        new_value.push_str(&value[last_match..m.start()]);
+        let placeholder_name = &caps[1];
+        let replacement = replacements.get(placeholder_name).ok_or(format!(
+            "Unknown value for placeholder `{}`",
+            placeholder_name
+        ))?;
+        new_value.push_str(replacement);
+        last_match = m.end();
+    }
+    new_value.push_str(&value[last_match..]);
+    // A replacement string may itself contain a placeholder, so we recurse.
```

Comment:
Hm... is this a feature that we already claimed to support? If not, it doesn't seem useful to support, and might possibly be confusing.

Reply:
We didn't claim it in the pantsbuild.org documentation (which was lacking other things too), but we did implement it, and document it in the Python code. I'm not keen on changing the behavior in what is supposed to be an invisible reimplementation.